published/external/xdpapi_helper.h

```diff
@@ -0,0 +1,289 @@
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+//
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <xdpapi.h>
+
+#if defined(_KERNEL_MODE)
+
+static const XDP_API_CLIENT _XdpApiContext = {0};
+static const XDP_API_PROVIDER_DISPATCH *_XdpApi;
+static const XDP_API_PROVIDER_BINDING_CONTEXT *_ProviderBindingContext;
+
+#define API_ABSTRACTION(FunctionCall, ...)  \
+        _XdpApi->FunctionCall((XDP_API_PROVIDER_BINDING_CONTEXT *)_ProviderBindingContext, \
+                              (XDP_API_CLIENT *)&_XdpApiContext, __VA_ARGS__)
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+NTSTATUS
+XdpHlpOpenApi(
+    _In_ UINT32 _XdpApiVersion,
+    _In_opt_ VOID *_ClientContext,
+    _In_opt_ XDP_API_ATTACH_FN *_ClientAttach,
+    _In_opt_ XDP_API_DETACH_FN *_ClientDetach,
+    _In_ const XDP_API_CLIENT_DISPATCH *_XdpApiClientDispatch,
+    _In_ const INT64 *_TimeoutMs
+    )
+{
+    return
+        XdpOpenApi(
+            _XdpApiVersion,
+            _ClientContext,
+            _ClientAttach,
+            _ClientDetach,
+            _XdpApiClientDispatch,
+            _TimeoutMs,
+            (XDP_API_CLIENT *)&_XdpApiContext,
+            (XDP_API_PROVIDER_DISPATCH **)&_XdpApi,
+            (XDP_API_PROVIDER_BINDING_CONTEXT **)&_ProviderBindingContext);
+}
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID
+XdpHlpCloseApi(
+    VOID
+    )
+{
+    XdpUnloadApi((XDP_API_CLIENT *)&_XdpApiContext);
+}
+
+inline
+_IRQL_requires_(PASSIVE_LEVEL)
+VOID*
+XdpHlpGetRoutine(
+    _In_z_ const CHAR* RoutineName
+    )
+{
+    if (!ExAcquireRundownProtectionCacheAware(_XdpApiContext.RundownRef)) {
+        return NULL;
+    }
+    VOID* Routine = _XdpApi->XdpGetRoutine(RoutineName);
+    ExReleaseRundownProtectionCacheAware(_XdpApiContext.RundownRef);
```

Comment:
Once this reference is released, how does the caller know it is safe to directly call the returned function?

Reply:
`XdpCreateProgram`, `XdpInterfaceOpen`, `XskCreate` and `XdpCloseHandleKernel` internally manages the reference and the object made by these call doesn't need ref count.
If I understand correctly, only this `XdpGetRoutine` internally need the reference count, not this helper level, right?